src/style/style_layer/fill_extrusion_style_layer.js

```diff
@@ -220,4 +253,96 @@ function projectExtrusion(geometry: Array<Array<Point>>, zBase: number, zTop: nu
     return [projectedBase, projectedTop];
 }
 
+/**
+ * Projects a fill extrusion vertices to screen while accounting for terrain.
+ * This and its dependent functions are ported directly from `fill_extrusion.vertex.glsl`
+ * with a few co-ordinate space differences.
+ *
+ * - Matrix `m` projects to screen-pixel space instead of to gl-coordinates (NDC)
+ * - Texture querying is performed in texture pixel coordinates instead of  normalized uv coordinates.
+ * - Height offset calculation for fill-extrusion-base is offset with -1 instead of -5 to prevent underground picking.
+ */
+function projectExtrusion3D(geometry: Array<Array<Point>>, zBase: number, zTop: number, translation: Point, m: Float32Array, demSampler: DEMSampler, centroid: vec2, exaggeration: number, lat: number) {
+    const projectedBase = [];
+    const projectedTop = [];
+
+    for (const r of geometry) {
+        const ringBase = [];
+        const ringTop = [];
+        for (const p of r) {
+            const x = p.x + translation.x;
+            const y = p.y + translation.y;
+            const heightOffset = getTerrainHeightOffset(x, y, zBase, zTop, demSampler, centroid, exaggeration, lat);
+
+            const base = toPoint(vec3.transformMat4([], [x, y, heightOffset.base], m));
+            const top = toPoint(vec3.transformMat4([], [x, y, heightOffset.top], m));
+
+            ringBase.push(base);
+            ringTop.push(top);
+        }
+        projectedBase.push(ringBase);
+        projectedTop.push(ringTop);
+    }
+    return [projectedBase, projectedTop];
+}
+
+function toPoint(v: vec3): Point {
+    const p = new Point(v[0], v[1]);
+    p.z = v[2];
+    return p;
+}
+
+function getTerrainHeightOffset(x: number, y: number, zBase: number, zTop: number, demSampler: DEMSampler, centroid: vec2, exaggeration: number, lat: number): { base: number, top: number} {
+    const ele = exaggeration * demSampler.getElevationAt(x, y, true, true);
+    const flatRoof = centroid[0] !== 0;
+    const centroidElevation = flatRoof ? centroid[1] === 0 ? exaggeration * elevationFromUint16(centroid[0]) : exaggeration * flatElevation(demSampler, centroid, lat) : ele;
+    return {
+        base: ele + (zBase === 0) ? -1 : zBase, // Use -1 instead of -5 in shader to prevent picking underground
+        top: flatRoof ? Math.max(centroidElevation + zTop, ele + zBase + 2) : ele + zTop
+    };
+}
+
+function elevationFromUint16(n: number): number {
+    return n / 7.3;
+}
+
+function flatElevation(demSampler: DEMSampler, centroid: vec2, lat: number): number {
+    const pos = [Math.floor(centroid[0] / 8), Math.floor(centroid[1] / 8)];
+    const span = [10 * (centroid[0] - pos[0] * 8), 10 * (centroid[1] - pos[1] * 8)];
```

Comment:
Where is this 10 coming from?
And why does pos need to be rounded to the nearest 8?